### PR TITLE
[TRTLLM-6393][feat] add static tree sampling and verification

### PR DESCRIPTION
**Pull request by yweng0828**
## Summary by CodeRabbit

* **New Features**
  * Optional static-tree speculative decoding for EAGLE3 via a new --eagle_choices CLI option or config; runtime parsing/validation and automatic adjustments to decoding settings and total draft-token limits.
  * Resource-manager-driven SpecTree integration enabling a tree-based drafting path; sampling and drafting flows now accept and propagate a resource manager when available.

* **Tests**
  * New unit tests for static-tree sampling and draft-token verification across layers, batches and large trees.

## Description

This RP implements sampling and verification of the draft token static tree.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/llmapi/llm_args.py [hunk 1]
```diff
@@ -425,6 +426,50 @@ class EagleDecodingConfig(DecodingBaseConfig):
     max_non_leaves_per_layer: Optional[int] = None
     eagle3_one_model: Optional[bool] = True
 
+    def __init__(self, **kwargs):
+        super().__init__()
+        for attr_name, attr_value in kwargs.items():
+            if attr_name == 'max_draft_len':
+                self.num_eagle_layers = attr_value
```
Comment (mikeiovine) [inline tensorrt_llm/llmapi/llm_args.py:433]: I feel like the user facing part of this can be refined a bit. I feel it would be most clear if `num_eagle_layers` and `max_draft_len` were the same. However, it seems like `max_draft_len` is being used as "the maximum number of total tokens" right now - not necessarily the maximum depth of the tree (correct me if I'm wrong about that). That's a bit confusing for readers of the code.

It would be great if we could have the following fields in the decoding config:
* `max_draft_len`: The maximum depth of the tree.
* `max_total_draft_tokens`: Should be the same as `max_draft_len` for chains. Can be larger for trees
* `num_eagle_layers`: We must keep it for backwards compatibility with TRT. But we can just say "do not use on new PyTorch code" in a comment. I feel we should also get rid of internal uses of `num_eagle_layers` and just use `max_draft_len` everywhere.

What do you think? Is it feasible to do like this?

File: tensorrt_llm/llmapi/llm_args.py [hunk 2]
```diff
@@ -425,6 +426,50 @@ class EagleDecodingConfig(DecodingBaseConfig):
...
+            # Convert the data type of Eagle choice from str to List[List[int]]
+            if attr_name == 'eagle_choices' and not isinstance(
+                    attr_value, list):
+                print(
+                    "NOTE: The Draft token tree is still under development, PLEASE DO NOT USE IT !!!"
+                )
+                if isinstance(attr_value, str):
+                    attr_value = ast.literal_eval(attr_value.replace(" ", ""))
+                else:
+                    assert "Wrong eagle choices type"
```
Comment (mikeiovine) [inline tensorrt_llm/llmapi/llm_args.py:443]: `assert` should take a bool. I think you meant to do this:

```
assert isinstance(attr_value, str), "Wrong eagle choices type"
```

Better yet, why not `raise ValueError()` here?

**Review by mikeiovine — COMMENTED** [review]
Great work!

**Review by mikeiovine — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/resource_manager.py [hunk 3]
```diff
@@ -1108,3 +1108,153 @@ def free_resources(self, request: LlmRequest):
 
     def shutdown(self):
         pass
+
+
+class SpecTreeManager:
```
[the remark this answers is not in the document]
Comment (yweng0828) [reply tensorrt_llm/_torch/pyexecutor/resource_manager.py:1129]: Thanks for the suggestion. I have moved the `SpecTreeManager` code to `spec_tree_manager.py` under `_torch/speculative`.

File: tensorrt_llm/_torch/speculative/spec_tree_manager.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+        # Initialize the buffers
+        self.eagle_paths = torch.ones(
+            (self.num_trees, self.max_total_draft_tokens + 1,
+             self.max_draft_len + 1),
+            dtype=torch.int32,
+            device='cpu',
+            pin_memory=True,
+        ) * -1
+        self.spec_dec_mask_matrix = torch.eye(
```
Comment (yweng0828) [inline tensorrt_llm/_torch/speculative/spec_tree_manager.py:49]: The main refactor is to refactor the tree information. Redundant data structures and calculations (such as parent nodes, child nodes, etc.) are removed. Get the tree information from the mask as we can. And we will keep updating the tree mask between different layers.

File: tensorrt_llm/_torch/speculative/spec_tree_manager.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        self.eagle_paths.fill_(-1)
+        self.eagle_paths[0][0][0] = 0  # root node
+        for i, choice in enumerate(self.eagle_choices):
+            self.eagle_paths[0][i + 1][0] = 0
+            for j in range(len(choice)):
+                self.eagle_paths[0][i + 1][j + 1] = index_mapping_set[str(
+                    choice[:j + 1])]
+
+        # 3) Compute node_list_per_layer
+        nodes_list_per_layer[0].append(0)  # root node
```
Comment (yweng0828) [inline tensorrt_llm/_torch/speculative/spec_tree_manager.py:98]: We may still need some auxiliary variables, but these are mainly for static trees because the structure of static trees is fixed. For dynamic trees, we need additional calculations/implementations (TODO).

File: tensorrt_llm/_torch/speculative/spec_tree_manager.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+        # 6) Compute the spec decoding according to the eagle_paths
+        for i, path in enumerate(self.eagle_paths[0]):
+            indices = path[path > -1]
+            self.spec_dec_mask_matrix[0][i, indices] = 1
+        self.spec_dec_pack_mask = self.compute_spec_dec_pack_mask(
+            self.spec_dec_mask_matrix)
+
+    # Get the eagle_paths
+    def get_eagle_paths(self, tree_idx=0):
```
Comment (yweng0828) [inline tensorrt_llm/_torch/speculative/spec_tree_manager.py:134]: The path is obtained from the mask and can be used for verification. For static trees, both the mask and the path are fixed. For dynamic trees, we need to dynamically update the mask and obtain the path based on the mask.

File: tests/unittest/_torch/speculative/test_draft_token_tree_sampling.py [hunk 7]
```diff
@@ -0,0 +1,415 @@
+import os
```
Comment (yweng0828) [inline tests/unittest/_torch/speculative/test_draft_token_tree_sampling.py:1]: The refactor has little impact on the implementation of test.

**Review by mikeiovine — APPROVED** [review]
Refactor looks good to me

**Review by Superjomn — APPROVED** [review]
LGTM on the llmapi changes

**Review by syuoni — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 8]
```diff
@@ -412,6 +437,16 @@ def get_generator(self, device: torch.device) -> torch.Generator:
             self._generator.manual_seed(self._global_seed)
         return self._generator
 
+    def get_spec_tree_manager(self, resource_manager: ResourceManager):
```
Comment (ixlmar) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:440]: `-> Optional[SpecTreeManager]`

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 9]
```diff
@@ -491,6 +526,94 @@ def _process_draft_tokens_greedy(self, request: LlmRequest,
                 break
         return num_accepted
 
+    def _process_draft_tokens_tree(self, request: LlmRequest,
+                                   new_tokens: torch.Tensor,
+                                   spec_tree_manager: SpecTreeManager) -> int:
+        """ Tree verification for draft token tree based speculative decoding. This function will only be called for the target model.
```
Comment (ixlmar) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:532]: Nit: Extraneous whitespace after `""" `

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 10]
```diff
@@ -533,15 +656,98 @@ def _process_draft_tokens_rejection_sampling(
...
+        assert len(topK_list) == num_logits_per_request
+
+        # TODO: this tensor can be optimized
+        new_draft_tokens_cuda = torch.zeros(
+            (len(requests), spec_tree_manager.max_total_draft_tokens + 1),
+            dtype=torch.int64,
+            device='cuda')
+        topK_list_cumsum = torch.cumsum(topK_list, dim=0)
+        # Different nodes have different topK value.
+        for i in range(len(topK_list)):
```
Comment (ixlmar) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:719]: * `for i, topK_list_i in enumerate(topK_list):`
* naming: `top_k_list` to conform to style guidelines

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 11]
```diff
@@ -533,15 +656,98 @@ def _process_draft_tokens_rejection_sampling(
...
+            assert logits.shape[0] == len(requests)
+            # 3) Sample the logits according to the topK value.
+            indices = torch.topk(logits, k=topK_list[i], dim=-1).indices
+            # 4) Write to the temporary output tensor.
+            new_draft_tokens_cuda[
+                request_index, topK_list_cumsum[i] -
+                topK_list[i]:topK_list_cumsum[i]] = indices[request_index]
+
+        # 5) Append eagle3 d2t.
+        self.append_eagle3(new_draft_tokens_cuda, model_outputs)
```
Comment (ixlmar) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:731]: Renamed to `_apply_d2t` in #7294

File: tensorrt_llm/llmapi/llm_args.py [hunk 12]
```diff
@@ -439,13 +440,69 @@ class EagleDecodingConfig(DecodingBaseConfig):
...
+        super().__init__()
+        for attr_name, attr_value in kwargs.items():
+            if attr_name == 'max_draft_len':
+                self.num_eagle_layers = attr_value
+                self.max_total_draft_tokens = attr_value  # If using linear-tree, the max_total_draft_tokens is the same as max_draft_len
+            # Convert the data type of Eagle choice from str to List[List[int]]
+            if attr_name == 'eagle_choices' and attr_value is not None:
+                logger.warning(
+                    "NOTE: The Draft token tree is still under development, PLEASE DO NOT USE IT !!!"
+                )
```
[the remark this answers is not in the document]
Comment (yweng0828) [reply tensorrt_llm/llmapi/llm_args.py:470]: Thanks for pointing this out. This is so we can call it during testing without throwing an error. That's why we only write a warning here. In places where the test doesn't execute but the user might execute, we write some `raise NotImplementedError`.

File: tensorrt_llm/_torch/pyexecutor/sampler.py (hunk 8 above)
Comment (yweng0828) [reply tensorrt_llm/_torch/pyexecutor/sampler.py]: Thanks, modified.

File: tensorrt_llm/_torch/pyexecutor/sampler.py (hunk 9 above)
Comment (yweng0828) [reply tensorrt_llm/_torch/pyexecutor/sampler.py]: Thanks, modified.

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 13]
```diff
@@ -533,15 +656,98 @@ def _process_draft_tokens_rejection_sampling(
...
+
+        # 1) Get the topK list for the specific draft layer.
+        topK_list = spec_tree_manager.get_topK_list(draft_layer_id)
+        assert len(topK_list) == num_logits_per_request
+
+        # TODO: this tensor can be optimized
+        new_draft_tokens_cuda = torch.zeros(
+            (len(requests), spec_tree_manager.max_total_draft_tokens + 1),
+            dtype=torch.int64,
+            device='cuda')
```
[the remark this answers is not in the document]
Comment (yweng0828) [reply tensorrt_llm/_torch/pyexecutor/sampler.py]: The new code does not use `new_tokens`, so I use `logits.device` here.

File: tensorrt_llm/_torch/pyexecutor/sampler.py (hunk 10 above)
Comment (yweng0828) [reply tensorrt_llm/_torch/pyexecutor/sampler.py]: Thanks for pointing it out. Both of them have been modified.

File: tensorrt_llm/_torch/pyexecutor/sampler.py (hunk 11 above)
Comment (yweng0828) [reply tensorrt_llm/_torch/pyexecutor/sampler.py]: Thanks, updated.

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 14]
```diff
@@ -533,15 +656,98 @@ def _process_draft_tokens_rejection_sampling(
...
+            torch.int, non_blocking=True)
+        int_new_draft_tokens = int_new_draft_tokens.unsqueeze(
+            dim=-1)  # [max_total_draft_tokens + 1, max_num_sequences, 1]
+        # new_tokens: device buffer, shape:[max_total_draft_tokens + 1, max_num_sequences, MAX_BEAM_WIDTH]
+        # new_tokens just store the new draft tokens for each layer.
+        new_tokens.index_copy_(1, seq_slots, int_new_draft_tokens)
+
+        if spec_tree_manager.use_dynamic_tree and draft_layer_id == spec_tree_manager.max_draft_len - 1:
+            # TODO: Re-sample the draft tokens for the last layer.
+            assert False, "Dynamic tree is not fully supported yet."
```
[the remark this answers is not in the document]
Comment (yweng0828) [reply tensorrt_llm/_torch/pyexecutor/sampler.py]: Thanks, updated.

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 15]
```diff
@@ -533,15 +656,98 @@ def _process_draft_tokens_rejection_sampling(
...
+            # 4) Write to the temporary output tensor.
+            new_draft_tokens_cuda[
+                request_index, topK_list_cumsum[i] -
+                topK_list[i]:topK_list_cumsum[i]] = indices[request_index]
+
+        # 5) Append eagle3 d2t.
+        self.append_eagle3(new_draft_tokens_cuda, model_outputs)
+
+        # 6) Copy back to the output tensor.
+        int_new_draft_tokens = new_draft_tokens_cuda.transpose(0, 1).to(
```
[the remark this answers is not in the document]
Comment (yweng0828) [reply tensorrt_llm/_torch/pyexecutor/sampler.py]: Yes, I tried redefining `new_draft_tokens_cuda` to `[max_total_draft_tokens + 1, max_num_sequences]`. However, `torch.topk` returns indices with shape `[num_sequences, top_k_list_i]`.

So, if we want to write from `indices [num_sequences, top_k_list_i]` to `new_draft_tokens_cuda [max_total_draft_tokens + 1, max_num_sequences]`, one tensor must be transposed.

**Review by kaiyux — APPROVED** [review]
Note sure why this PR requires approval from trt-llm-doc-owners, approving to unblock the merge.

**Review by suyoggupta — APPROVED** [review]
